Migrate to new protobuf implementation

michaelbausor opened the pull request:
This PR still needs some cleaning up, but I wanted to start getting review sooner rather than later. Some points:
- The most significant change is to logging, pubsub and spanner, because they rely on GAPICs. In particular, the `PhpArray` codec is replaced by a `Serializer` object.
- The huge number of files is because this PR adds all of the generated protobuf types - these don't need review individually, but we do want to consider what will be the best repo structure for including these types (if indeed we decide to go ahead with that plan), considering our release process.

dwsupplee commented on src/Logging/V2/ConfigServiceV2Client.php:
```diff
@@ -534,7 +544,7 @@ public function createSink($parent, $sink, $optionalArgs = [])
      * Updates a sink. If the named sink doesn't exist, then this method is
      * identical to
-     * [sinks.create](https://cloud.google.com/logging/docs/api/reference/rest/v2/projects.sinks/create).
+     * [sinks.create](/logging/docs/api/reference/rest/v2/projects.sinks/create).
```
We'll need to retain these links :)

michaelbausor replied:
Definitely, I will regenerate with these included

dwsupplee commented on src/proto/Google/Cloud/Language/V1beta2/AnalyzeEntitiesRequest.php:
```diff
@@ -0,0 +1,93 @@
+     * Input document.
+     * </pre>
+     *
+     * <code>.google.cloud.language.v1beta2.Document document = 1;</code>
```
Will we be able to have these generated with valid docblocks?

michaelbausor replied:
We have an open issue here: https://github.com/google/protobuf/issues/3118

dwsupplee commented on src/google-cloud-php.iml:
```diff
@@ -0,0 +1,9 @@
+<?xml version="1.0" encoding="UTF-8"?>
```
This shouldn't be committed

michaelbausor replied:
Removed

dwsupplee commented:
There are some types that still live in proto-client-php, for example:

https://github.com/googleapis/proto-client-php/tree/proto3/src/Google/Logging/Type

Should these be a part of this PR as well?

jdpedrie commented on src/Core/Serializer.php:
```diff
@@ -0,0 +1,203 @@
+                }
+                $v = $arr;
+            }
+            elseif ($field->isRepeated()) {
```
nit: this should be on the previous line, e.g. `} elseif () {`

michaelbausor replied:
Fixed

dwsupplee commented:
If we do keep the protos in this repository, what do you think of the following?

Instead of `Google\Logging\V2\CreateSinkRequest` - dir: `src/proto/Logging/V2`

`Google\Cloud\Logging\V2\CreateSinkRequest` - dir:  `src/Logging/V2`

A major downside here, is that the protos will live right next to the GAPIC clients - making it hard to distinguish which to use if looking directly at the source. An option then would be to use `Google\Cloud\Logging\V2\Proto\CreateSinkRequest` - dir: `src/Logging/V2/Proto`.

On the plus side, we stick to our core namespace of `Google\Cloud` and managing releases will be easier.

michaelbausor commented:
@dwsupplee That seems possible, but one problem is that the proto types do not all use the Google\Cloud namespace (e.g. the logging types are under `Google\Logging` rather than `Google\Cloud\Logging`). This is something that we could perhaps push for, by adding support for an option to specify the namespace in the proto file similar to C#: https://github.com/googleapis/googleapis/blob/6ae2d424245deeb34cf73c4f7aba31f1079bcc40/google/logging/v2/logging.proto#L28

dwsupplee commented:
That sounds like a great solution to me

michaelbausor commented:
Given that, I think it might be best if we make the migration to the new protobuf implementation **without** including the generated types in google-cloud-php, and instead leaving them in proto-client-php. Then once we have that protobuf feature available, we can change the namespaces and move the generated classes. This has the downside of involving two disruptive changes, but the upside of not blocking on having that protobuf feature available. WDYT?

dwsupplee commented:
Is it correct then that for packages we are currently GA on we would end up bumping them to 3.0 in a relatively short time frame? I have some definite reservations there - do you have an idea as to when it may be possible to get the update to change the namespaces?

michaelbausor commented:
No, we should not be bumping them beyond 1.0. The affected GA API is Logging, but in that case we have gone GA on the manual layer but not on the GAPIC client or protobuf types, because we knew that these breaking changes would be required (and we documented that the GAPIC layer is still experimental).

michaelbausor commented:
An update after some discussions with @bshaffer, @tmatsuo and @TeBoring:

We are now looking to implement the proto option to set the PHP namespace (issue filed here: https://github.com/google/protobuf/issues/3122) before we make the migration over to the new protobuf implementation. This will allow us to move all of the generated protobuf types into the `Google\Cloud` namespace.

There is also a new issue in protobuf that I discovered today while running Spanner integration tests, which we will need to resolve before we can complete the migration: https://github.com/google/protobuf/issues/3139

michaelbausor commented:
I have updated the PR to include:
- Protobuf types are now in src/Google/Cloud directories and have correct namespaces (generated from https://github.com/google/protobuf/pull/3162). I think that the clutter, while not ideal, it not too bad - README samples can direct users to the client class, and there is a single place to look if you want to find a class to use.
- The protobuf metadata classes are in src/GPBMetadata. I think this is the best place for them - they all reside under the `GPBMetadata\\` namespace, and should never be used directly, so having them in the src/Google/Cloud directories doesn't seem right.
- Moved the Serializer class into GAX. I think that the functionality will be useful in autogen code as well as google-cloud-php. WDYT?

This PR is awful to review because of the proto files. Since it is not passing test yet anyway, WDYT if I remove the proto types for now?

jdpedrie commented:
> The protobuf metadata classes are in src/GPBMetadata.

That directory will not be distributed to components. If we are keeping the metadata classes together, they should probably be in `src/Core`?

michaelbausor commented:
Yes, I wanted to discuss that point. I am not sure if we want to:
- Keep them in Core
- Distribute them to the components, by doing one of:
  - Have the release process know which classes to pluck from the combined GPBMetadata folder
  - Have them separated, with each API directory having a metadata folder (e.g. `src/PubSub/GPBMetadata`)

Which of those do you think is preferable? All else equal, I think that having them distributed is best, by neither of the options for achieving that is ideal.

dwsupplee commented:
With option 2 would it would look something like this?

`src/PubSub/V1/GPBMetadata/Pubsub.php`

If so, I think I like that approach best.

michaelbausor commented:
Having `src/PubSub/V1/GPBMetadata/Pubsub.php` seems fine to me.

The downside of this approach is the noise that it creates in the `composer.json` file, in the `PSR-4` section, because the namespace of these metadata files begins with `GPBMetadata`. So in this example, we would need to update the autoload section of the composer file with:

```

michaelbausor commented:
Having src/PubSub/V1/GPBMetadata/Pubsub.php seems fine to me.

The downside of this approach is the noise that it creates in the composer.json file, in the PSR-4 section, because the namespace of these metadata files begins with GPBMetadata. So in this example, we would need to update the autoload section of the composer file with:

```
  "autoload": {
        "psr-4": {
            "Google\\Cloud\\": "src",
            "GPBMetadata\\Google\\Pubsub\\V1": "src/PubSub/V1/GPBMetadata"
        }
    },
```

Is that acceptable do you think? Otherwise we could look at using classmap rather than psr-4 for these GPBMetadata classes.

jdpedrie commented:
In my opinion, having the desired directory structure and distributing the metadata classes along with the relevant package is worth the extra noise in the composer.json file. :)

dwsupplee commented:
Will it not be possible to modify the namespace of the GPBMetadata classes?

michaelbausor commented:
No, unfortunately not. I mentioned this briefly in the PR for adding the php_namespace option (https://github.com/google/protobuf/pull/3162) but that was referring to changing the namespace to (in the pubsub example) include the `Cloud` component - not to remove the leading `GPBMetadata`, which is the cause of the noise.

We can clean up the composer file by placing them in a shared GPBMetadata folder in the root or `src`, but then we would need to have our release process pluck the correct files out, and distribute them to a GPBMetadata folder in each subrepo.

dwsupplee commented:
Gotcha. I'm a bit concerned about the room for err we would be introducing by having to remember to add mappings to the composer's autoload settings for every new library. It isn't the end of the world, but just one more thing that would be easy to forget.

Ultimately, if we keep everything in a shared directory, does it sound correct we would need a mapping for the split process? If so, it appears we are at the same problem (with having to maintain a map), just in a different place.

Between the two, I would prefer with maintaining the mapping in the composer's autoload setting just because it might be a more clear place to manage the information.

dwsupplee commented:
> Moved the Serializer class into GAX. I think that the functionality will be useful in autogen code as well as google-cloud-php. WDYT?

👍 sounds good to me.

jdpedrie commented:
I agree that between the two, the autoloader is preferable. I'd rather not introduce an opportunity for failure which is hard to catch before release, and if we use the autoloader, there is less room for surprise when the component acts entirely differently from the main package with respect to the metadata.

michaelbausor commented:
@dwsupplee you are correct that we would need to maintain a mapping in either case.

Sounds like there is agreement to use the autoloader to hold the mapping. If we decide in future that this is not working, we can restructure, which should be non-breaking to users.

michaelbausor commented:
Moved the GPBMetadata classes, and added to composer.json - using `classmap` seemed the cleanest way to include them.

I will also update each API-specific composer file.

michaelbausor commented:
> There are some types that still live in proto-client-php, for example:
>
>https://github.com/googleapis/proto-client-php/tree/proto3/src/Google/Logging/Type
>
>Should these be a part of this PR as well?

These common types will live in GAX, as they did previously (e.g. the existing HttpRequest type is here: https://github.com/googleapis/gax-php/blob/master/src/generated/http_request.php)

I should not have added them to the proto3 branch in the first place

dwsupplee commented:
> Moved the GPBMetadata classes, and added to composer.json - using classmap seemed the cleanest way to include them.

Just wanted to note that with a classmap new classes added to the directories won't be autoloaded unless we explicitly run composer install/update/dump-autoload.

I'm betting it will be extremely rare that a new class will be generated in the existing directories, but in the case they do it could lead to some confusion for anyone contributing to the project.

michaelbausor commented:
I am happy to not use classmap if it may cause issues, but I think you are right that it will be extremely rare - it will only occur when someone is regenerating the protobuf types for an API, and a new proto _file_ has been added to the API, because each GPBMetadata class corresponds to a file, rather than a proto message.

dwsupplee commented:
> These common types will live in GAX, as they did previously (e.g. the existing HttpRequest type is here: https://github.com/googleapis/gax-php/blob/master/src/generated/http_request.php)

Sounds good! Does it seem worthwhile to have portions of GAX run through the documentation generator so we can include these types on the site easily?

> I am happy to not use classmap if it may cause issues, but I think you are right that it will be extremely rare - it will only occur when someone is regenerating the protobuf types for an API, and a new proto file has been added to the API, because each GPBMetadata class corresponds to a file, rather than a proto message.

I'll defer to your judgement on this one :). I just wanted to make sure it was something we were aware of.

michaelbausor commented:
>Does it seem worthwhile to have portions of GAX run through the documentation generator so we can include these types on the site easily?

Good question, I think that makes sense, but I'm not 100% sure. Is that the best way to integrate the docs?

dwsupplee commented:
The other option I can think of is to do what we are doing currently - which is to just link out to the code in the repo. Having those types be documented in the site seems like it would be more cohesive.

Are there any other existing plans to document GAX currently?

michaelbausor commented:
No, there isn't any specific plan, other than generating a github.io doc site for GAX. Linking out to that would be OK, but less cohesive as you mention.

michaelbausor commented:
It looks like tests are very close to passing, there are a couple of issues:
- Some of the comments in the protobuf types are being picked up as snippets, as they contain "Example:". We could resolve this by (a) tightening the rules around what counts as a snippet, or (b) perhaps adding a new `@snippet` or `@example` annotation, or (c) excluding the V* directories from the snippet search.
- Doc gen is failing because protobuf type constructors don't have doc blocks. Ideally we would solve this by adding doc blocks to them - is there any other plausible way forward in the short term?

dwsupplee commented:
- Does updating "Example:" to "Sample Code:" sound okay? I originally didn't use the [`@example`](https://phpdoc.org/docs/latest/references/phpdoc/tags/example.html) tag because it doesn't support inline examples.
- The quickest route I can think of would be to add a condition during doc generation to exclude a certain pattern. Are we planning to build the docs for the protobuf types using the docblocks that currently exist? They don't use proper notation and will look pretty different from the rest of the docs on the site, at times even rather confusing.

michaelbausor commented:
- Moving to "Sample Code:" SGTM
- We are going to improve the generated docblocks - @bshaffer has a PR open here: https://github.com/google/protobuf/pull/3131. I can regenerate using that PR and we can assess the state of the docs and discuss what additional changes we want. But I think that we should try not to delay switching to the new protobuf - if the doc situation is not acceptable, then in the worst case we can leave the protobuf types in proto-client-php (but still make use of the updated namespace option, so that moving them from proto-client-php to google-cloud-php will be a non-breaking change)

michaelbausor commented:
I have added support in the documentation generator for classes and methods without doc blocks. This is not something that I want to encourage/make normal though. Currently the places that are missing doc blocks are the protobuf type constructors and the gRPC stub classes, which have comments but use `//` instead of `/* ... */`, so those are the changes we need to get from protobuf and grpc respectively.

The question for us I think is - does moving to this new distribution (protobuf types in google-cloud-php) improve our doc situation (vs the current option of linking out to code) or make it worse, given that the generated docs are not fully finished?

dwsupplee commented:
If these changes to doc generation will be here for some time, could we add a warning if there isn't a docblock? That will provide at least some coverage to make sure we don't miss something when developing new code.

[Here](https://dwsupplee.github.io/gcloud-php/#/docs/google-cloud/master/logging/v2/logentry) is an example of what the generated docs will look like with the new protobuf types. This will require a lot of manual work for a user to grok what is going on and to be able to link out to the different types required to build up a message. 

If we are really pressed to get this out before we can get the updated docs, then I don't see a difference between doing what we are currently (linking out to the code) vs generating into the doc site since the source of the data provides the same information.

michaelbausor commented:
I think that we need to include the param and return type updates from https://github.com/google/protobuf/pull/3131. I will regenerate soon with those changes, and we can include them even if they are not in a release of protobuf. I don't know if we are going to be pressed to get the switch out before we can make the updates to the doc gen to fix the missing docblocks - it depends on when protobuf can provide us with a new point release to depend on (which we need for the runtime changes, not the doc changes). But I do want to try as hard as possible not to be blocked once that is available.

Agree that if we include those changes to allow empty doc blocks, we should add warnings.

It looks like tests are passing, except for HHVM, because of this issue: https://github.com/google/protobuf/issues/3224
HHVM isn't listed as supported in the [PHP protobuf requirements readme](https://github.com/google/protobuf/tree/master/php#requirements) - I will ping the issue to find out what the intention is about HHVM support.

dwsupplee commented on src/ErrorReporting/V1beta1/ErrorGroupServiceGrpcClient.php:
```diff
@@ -0,0 +1,64 @@
+// See the License for the specific language governing permissions and
+// limitations under the License.
+//
+namespace Google\Cloud\ErrorReporting\V1beta1 {
```
There are a few oddities with this class and others like it:

Why the bracketed namespace?
2 spaces instead of 4?

michaelbausor replied:
The Grpc Client classes are coming from different generation code - the grpc plugin rather than protobuf. I will look at what is required to make them consistent, including fixing the doc block issue.

michaelbausor replied:
Filed a PR here: https://github.com/grpc/grpc/pull/11524

michaelbausor commented:
Updated protobuf and grpc type docs - PTAL and let me know if there are other doc generation problems.

dwsupplee commented:
A few items of note:

- There is a lack of consistency with how we are notating required items. In some cases this information provides no benefit, as it is clear an item is required if it is not optional in the signature. In other cases, I believe it is intended to let a user know which properties on a message must be set before the message can be delivered successfully, this makes more sense. Examples: [1](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/files#diff-7f27caeaa234f9174f9756c3dd215613R22), [2](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/files#diff-fcc8529918730286528eb5cb38e9fca5R23), [3](https://github.com/michaelbausor/gcloud-php/blob/b3f6eaf82cc85a738be72d3ec1b3cb6a9ff97fb5/src/Speech/V1/SpeechClient.php#L308)
- Are the `<pre>` tags necessary? They screw with the styling on the doc site, and I don't see the value in them being there.
- Should we be keeping the [code samples](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/files#diff-7f27caeaa234f9174f9756c3dd215613R64)? Ex: `<code>string project_name = 1;</code>`. I may be missing something, but I don't see how this will benefit a user - it just seems like it would mislead them.

michaelbausor commented:
- This is a consequence of the proto documentation. APIs have different conventions, and our documentation for the protobuf types and the GAPIC clients is just copying their comments. See https://github.com/googleapis/googleapis/blob/master/google/devtools/clouderrorreporting/v1beta1/error_stats_service.proto#L289 and https://github.com/googleapis/googleapis/blob/master/google/cloud/speech/v1/cloud_speech.proto#L66. An example of how this SHOULD be done at the proto level (IMHO) is pubsub: Use the proto comments on the message fields to describe the field (https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub.proto#L324) and document on the RPC the behaviour of whether fields are required, or what happens if they are missing (https://github.com/googleapis/googleapis/blob/master/google/pubsub/v1/pubsub.proto#L46). I don't think there is much we can do about this in autogen - as we would need to process the comments on a case-by-case basis.
- I don't see why they are necessary, lets look at removing them
- I think this is just supposed to document the underlying protobuf field from which the method was generated. It should probably not use `<code>` tags, lets look at changing them

michaelbausor commented:
Alternatively for the `<code>` tags, we could just make it clearer what they represent by changing:
`<code>string project_name = 1;</code>`
to
`Generated from protobuf message field: <code>string project_name = 1;</code>`
which is more similar to the docblock on the class: https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/files#diff-7f27caeaa234f9174f9756c3dd215613R16

dwsupplee commented:
I like that updated messaging better, but still would prefer removing it altogether. What is a user going to do with that information?

michaelbausor commented:
I our case it is probably not that useful, but in general for protobuf generated types it is there to indicate which proto field it came from.

I don't think they will want to remove it entirely. We could perhaps filter it from our doc site, if we thought it was worth it? But I would be unsure about if it was worth handling as a special case.

michaelbausor commented:
Regarding documentation of protobuf classes if they are not included in google-cloud-php, what do you think about:
-  Document them in the main doc site. I don't know what structure we would use though - any ideas?
- Or, use a separate doc site. Based on what I have looked at before, something like this: https://michaelbausor.github.io/proto-client-php/master/

michaelbausor commented:
@dwsupplee @jdpedrie Protobuf has [released 3.3.2](https://github.com/google/protobuf/releases/tag/v3.3.2), so we are able to switch to using the new protobuf implementation as soon as we are ready. Tests are now passing in this PR. The things that I think need review/consideration are:
- I have updated Core so that there are not breaking changes, by adding GPBGrpcTrait and similar for Grpc-based classes that need to change, and adding deprecation notices to GrpcTrait. Do you think this is a reasonable approach? When Core goes to v2, we would remove the GrpcTrait and rename GPBGrpcTrait to GrpcTrait. Note that I have also retained `stanley-cheung/protobuf-php` as a dev dependency - this was needed 
- I did a similar thing with `Logging/Connection/Grpc` , which is the other file in a GA component affected (excluding the Logging GAPIC clients)
- What are your thoughts about handling docs for proto-client-php (see previous comment)?
- I removed testing of HHVM from Travis - this won't be supported by protobuf until https://github.com/google/protobuf/pull/3225 is included in a release. Should I instead disable only the grpc tests somehow?
- Other concerns?

dwsupplee commented:
> I have updated Core so that there are not breaking changes, by adding GPBGrpcTrait and similar for Grpc-based classes

Since the public methods in these traits haven't changed their signature/return value and the rest of the methods are private - I think it would be acceptable to continue using them as is with the updates we require. The only class we should need to deprecate would be `Google\Cloud\Core\PhpArray`, I believe.

> Note that I have also retained stanley-cheung/protobuf-php as a dev dependency - this was needed to support tests passing, in particular the snippet test suite, which otherwise complains that DrSlump\Protobuf\Codec\PhpArray is missing.

Interesting 🤔, which test(s) did this crop up on?

> What are your thoughts about handling docs for proto-client-php?

For now, I like the idea of using our current structure to link out to the generated types hosted here: https://michaelbausor.github.io/proto-client-php. Once we get everything set up in the ideal state, we could then move forward with generating them right into the main docs site - WDYT?

> I removed testing of HHVM from Travis - this won't be supported by protobuf until google/protobuf#3225 is included in a release. Should I instead disable only the grpc tests somehow?

I would definitely prefer to disable the grpc tests in HHVM vs. disabling HHVM coverage altogether. We do something similar already by [checking if the grpc extension is installed](https://github.com/GoogleCloudPlatform/google-cloud-php/blob/master/tests/unit/Logging/Connection/GrpcTest.php#L38-L40).

```
if (!defined('HHVM_VERSION')) {
    $this->markTestSkipped('HHVM does not currently support gRPC.');
}
```

> Other concerns?

None at the moment! I will spend some more time tonight looking through the PR to confirm.

michaelbausor commented:
> Since the public methods in these traits haven't changed their signature/return value and the rest of the methods are private - I think it would be acceptable to continue using them as is with the updates we require. The only class we should need to deprecate would be Google\Cloud\Core\PhpArray, I believe.

I am not opposed to this necessarily, but wouldn't this still count as a breaking change? We would be changing the types of callables that are accepted - callables that use the DrSlump classes would no longer work.

> Interesting :thinking:, which test(s) did this crop up on?

It happens in the parsing step before running any snippet tests. I will investigate further.

> For now, I like the idea of using our current structure to link out to the generated types hosted here: https://michaelbausor.github.io/proto-client-php. Once we get everything set up in the ideal state, we could then move forward with generating them right into the main docs site - WDYT?

+1

> I would definitely prefer to disable the grpc tests in HHVM vs. disabling HHVM coverage altogether. We do something similar already by checking if the grpc extension is installed.

I will investigate doing this

dwsupplee commented:
> I am not opposed to this necessarily, but wouldn't this still count as a breaking change? We would be changing the types of callables that are accepted - callables that use the DrSlump classes would no longer work.

Good point, after thinking about it a bit I am leaning towards using the classes as is. The breaking change technically takes place in parts of the library we have marked as experimental, whereas this is just a generic wrapper. WDYT @jdpedrie?

michaelbausor commented:
>> Interesting :thinking:, which test(s) did this crop up on?

> It happens in the parsing step before running any snippet tests. I will investigate further.

ReflectionClass throws an error if a class cannot be instantiated, which happens in the case of `PhpArray` if it doesn't have the `DrSlump` dependency. I could update the process to:

- Skip deprecated files/classes. But this is tricky without using reflection...
- Have an explicit skip list
- Keep the dev dependency
- Other?

dwsupplee commented:
> Keep the dev dependency

I'm okay with this

michaelbausor commented:
After much struggling with HHVM, I have come up against a tricky problem. The reason it is failing is [this code](https://github.com/google/protobuf/blob/v3.3.2/php/src/Google/Protobuf/descriptor.php#L84) in protobuf. This is problematic because (1) this code is a PARSE-time error in HHVM, not a runtime error, and (2) this code does NOT follow PSR-4, meaning it is not autoloaded, but always loaded. Therefore, just skipping the gRPC tests is not sufficient - it is necessary to conditionally remove the dependencies on google/protobuf, google/gax and google/proto-client-php for HHVM.

After doing that, we then hit the problem on the snippet parser, which also tries to load all the classes and errors when it can't.

I think this can be solved by adding an exclude list for the snippet parser, which checks for HHVM and conditionally does not check the Spanner files for snippets. This would also allow use to remove the stanley-cheung/protobuf-php dependency.

dwsupplee commented:
That sounds like a solid solution for now, and hopefully we can get HHVM support sooner rather than later.

michaelbausor commented:
HHVM tests are now passing. I did not need to use an exclusion list for grpc snippets in the end, but having implemented it, I have retained it to remove the `DrSlump` dev dependency.

I believe this is ready for final review, pending releases of GAX and proto-client-php, and updating their docs to link externally instead of to code.

dwsupplee commented on src/Spanner/composer.json:
```diff
@@ -5,9 +5,9 @@
-        "google/cloud-core": "^1.4",
-        "google/gax": "^0.10",
-        "google/proto-client-php": "^0.13"
+        "google/cloud-core": "^1.6",
```
To allow travis to pass (since it is running a check to determine if components can be installed properly) we will need to relax this to a currently tagged release.

michaelbausor replied:
Done - although this seems like a potential problem in our release process, if we cannot do a simultaneous release of two components, one of which depends on the new release of the other. It may be that this is simply a problem of composer, where in order to install a local package it is necessary to modify the composer file. We can discuss and try to resolve this outside of this PR.

dwsupplee commented:
Did we need to revert the snippet changes (https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/commits/9c4b04bdbff31f26a8f7146fcb51ae8860f958dc) (https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/commits/a518c928bc41e642854b8000bd43dfcf5312dbb2)?

michaelbausor commented on dev/src/Snippet/Coverage/Coverage.php:
```diff
@@ -17,8 +17,14 @@
 
 namespace Google\Cloud\Dev\Snippet\Coverage;
 
+use Google\Cloud\Dev\CheckGrpcTrait;
```
Good point, removed the trait and moved `shouldSkipGrpcTests` to `tests/GrpcTestTrait.php`

michaelbausor commented on docs/external-classes.json:
```diff
@@ -1,98 +1,50 @@
 }, {
   "name": "Google\\GAX\\",
-  "uri": "https://github.com/googleapis/gax-php/tree/master/src/%s.php"
+  "uri": "https://googleapis.github.io/gax-php/master/Google/GAX/%s.html"
```
I am happy to do that, updated.

My only concern is that we will need to keep this updated as new docs are published. I am happy to try pointing at the most recent docs, and if we find problems keeping everything updated we can try a different approach.

dwsupplee replied:
In the docs site we use on google-cloud-* we have an alias `latest` that will always point to the most recent release. Maybe Sami has something similar?

jdpedrie replied:
Perhaps the doc generator could parse `composer.json` and get the correct version of gax to inject into the URL?

michaelbausor replied:
Hmm good point. I have just pushed a change so that `#` is used to redirect to the latest docs, but maybe this isn't what we want.

Parsing the composer file could work, but pinning to a specific GAX version doesn't seem quite right either. If our dev dependency is `"google/gax": "^0.20.0"` but we release a `0.20.1` version, we would probably want to point to that instead.

Similarly with the auth library, we depend on `^1.0` so it seems to make sense to point to `1.1` if it is available, because that is what a new user will get if they run `composer install`

michaelbausor replied:
After some thought, I think that we have three cases: `x.y.z` (specific-version), `^1.x` (GA), and `^0.x` (pre-GA), which actually work out very similar in practice. For a specific-version dependency, we just point to the docs for that specific version. For a `^1.x` dependency, we want to point to the latest `1.*` release. In practice, this is the auth library, and pointing at master makes sense - any changes will be driven by our teams and included in a release, and we can coordinate a 2.0 release if that is ever necessary. For a `^0.x` dependency, we want to point to the latest `0.x.*` release. In practice, this is the GAX and proto-client libraries, which we control. Whenever we release a new version, it will be because we want to use it in google-cloud-php, and we can control the GAX doc site redirection so that it gets updated only when google-cloud-php takes a dependency on the latest release (see https://github.com/googleapis/gax-php/blob/gh-pages/index.html). Once GAX goes GA, the situation will be similar to the auth library, and we will want to point to the latest release all of the time. Similarly we can also coordinate a 2.0 release with google-cloud-php so that the docs are not out of sync if that is ever necessary.

So, the upshot is that I think leaving the docs with the current behaviour (redirecting to "latest" for GAX, pointing to master for auth) might be the way to go.

WDYT?

dwsupplee replied:
Sounds like a pretty fair compromise. Works for me!

dwsupplee commented:
> We didn't need to, but because we aren't including the protobuf classes, there wasn't a reason to change them, and I wanted to minimize the changes in this PR. If you think they are a worthwhile change in their own right (or just worth adding as preparation) and can add them back in.

I wasn't sure if lines [like this](https://github.com/GoogleCloudPlatform/google-cloud-php/pull/512/files#diff-01a51c98c3a1e1f31de2a3c61c575f33L418) may cause issues, but after taking a quick look at the docs it appears we should be okay. Given that, I'm happy to keep the changes in this PR minimal, that works for me!

michaelbausor commented:
Great! I believe I have addressed all changes, so this should be ready to merge - PTAL
